cmake/Modules/FindNCCL.cmake

```diff
@@ -14,32 +14,34 @@
 # install NCCL in the same location as the CUDA toolkit.
 # See https://github.com/caffe2/caffe2/issues/1601
 
-set(NCCL_ROOT_DIR $ENV{NCCL_ROOT_DIR} CACHE PATH "Folder contains NVIDIA NCCL")
+set(NCCL_INCLUDE_DIR $ENV{NCCL_INCLUDE_DIR} CACHE PATH "Folder contains NVIDIA NCCL headers")
+set(NCCL_LIB_DIR $ENV{NCCL_LIB_DIR} CACHE PATH "Folder contains NVIDIA NCCL libraries")
+set(NCCL_VERSION $ENV{NCCL_VERSION} CACHE STRING "Version of NCCL to build with")
+
+list(APPEND NCCL_ROOT ${NCCL_ROOT_DIR} ${CUDA_TOOLKIT_ROOT_DIR})
+# Compatible layer for CMake <3.12. NCCL_ROOT will be accounted in for searching paths and libraries for CMake >=3.12.
+list(APPEND CMAKE_PREFIX_PATH ${NCCL_ROOT})
 
 find_path(NCCL_INCLUDE_DIRS
   NAMES nccl.h
-  HINTS
-  ${NCCL_INCLUDE_DIR}
-  ${NCCL_ROOT_DIR}
-  ${NCCL_ROOT_DIR}/include
-  ${CUDA_TOOLKIT_ROOT_DIR}/include)
+  HINTS ${NCCL_INCLUDE_DIR})
 
-IF ($ENV{USE_STATIC_NCCL})
-  MESSAGE(STATUS "USE_STATIC_NCCL detected. Linking against static NCCL library")
-  SET(NCCL_LIBNAME "libnccl_static.a")
-ELSE()
+if (USE_STATIC_NCCL)
+  MESSAGE(STATUS "USE_STATIC_NCCL is set. Linking with static NCCL library.")
+  SET(NCCL_LIBNAME "nccl_static")
+  if (NCCL_VERSION)  # Prefer the versioned library if a specific NCCL version is specified
+    set(CMAKE_FIND_LIBRARY_SUFFIXES ".a.${NCCL_VERSION}" ${CMAKE_FIND_LIBRARY_SUFFIXES})
+  endif()
+else()
   SET(NCCL_LIBNAME "nccl")
-ENDIF()
+  if (NCCL_VERSION)  # Prefer the versioned library if a specific NCCL version is specified
+    set(CMAKE_FIND_LIBRARY_SUFFIXES ".so.${NCCL_VERSION}" ${CMAKE_FIND_LIBRARY_SUFFIXES})
```

Comment:
I'm confused about the placement of this line. It occurs inside the `USE_STATIC_NCCL` block. Doesn't that mean that it will never actually usefully trigger (because we'll be linking against static nccl?)

Reply:
I think this is under the `else`-clause? This line will be triggered only if `USE_STATIC_NCCL` is false.